Board Support for Sparkfun WRL13990 (nrf52832)

rolandvs opened the pull request:
The WRL13990 is a minimum development board from Sparkfun.

It features an nRF52832 and a minimum of extra's like a button and a LED. The UART connector has the FTDI 6-pin format for easy access using the USB-RS232 cable of FTDI.

rolandvs committed:
Board Support for Sparkfun WRL13990 (nrf52832)

Board support for this module.

aykevl commented on ports/nrf/boards/WRL13990/mpconfigboard.h:
```diff
@@ -0,0 +1,82 @@
+// Connect Macronix MX25R4035F (4MBit) or MX25R8035F (8Mbit)
+#define MICROPY_HW_SPI1_NAME    	"SPI1"
+#define MICROPY_HW_SPI1_MISO		(20)
+#define MICROPY_HW_SPI1_CLK			(16)
```
I see you have been using tabs here. MicroPython uses spaces for indentation.

aykevl commented on ports/nrf/boards/WRL13990/mpconfigboard.h:
```diff
@@ -0,0 +1,82 @@
+
+// options to check out
+//#define MICROPY_PY_BLE				(0)
+//#define MICROPY_HW_HAS_SDCARD		(0)
```
This code (and other commented out code below) isn't used. Can you remove it, just for cleanliness?

Last time I tested SD card support using this flag doesn't work. `sdcard.py` probably works better.

aykevl reviewed (changes requested):
Thank you for the contribution! I did a really quick review.

aykevl commented:
Can you also add this board to the README?

rolandvs committed:
Added port to the readme

rolandvs committed:
Removed tabs and unused definitions

glennrub commented on ports/nrf/boards/WRL13990/mpconfigboard.h:
```diff
@@ -0,0 +1,74 @@
+#define MICROPY_HW_SPI0_MOSI        (23)
+#define MICROPY_HW_SPI0_MISO        (24)
+
+// SPI1 config
```
The defines for SPI1 would not have any effect as the machine/spi.c will only render SPI0 compile time. Does the Sparkfun breakout boards have a implicit board definition of where SPI1 is expected to be located? Like Arduino board layout expects SPI to be on pins located at the position of 11-13 and often CS on 10. If so, it could be an option to also update the modules/machine/spi.c to interpret a second compile time SPI. 

If there is no expectation on where SPI1 should be on the board i would probably have dropped this from the board definition as this would vary quite a lot between use cases of the board and would also require to implement the necessary hooks in the machine/spi.c to make them used. You can always define the pins used for mosi, miso, sck and cs from the machine.SPI module.

rolandvs replied:
@glennrub - thanks for your remark. I will change it as indicated. There is no "assigned" SPI1 whatsoever so lets make it as simple as possible, but not simpler as a famous person once said :-).

rolandvs committed:
Removed all unnecessary defines to make it as clean as possible

glennrub commented on ports/nrf/boards/WRL13990/pins.csv:
```diff
@@ -0,0 +1,52 @@
+P29,P29
+P30,P30
+P31,P31
+BTN,P6
```
Could it be an option to replace the template definitions on top with the new functionality names and get them listed in ascending order? 

Also, if there is a need for fixing the position of SPI1 it would be good to align the names between SPI0 and SPI1, CLK and SCK.

rolandvs replied:
I would like to keep the generic names P0..Pxx as well as the labeled pins. Maybe just move all the labeled pins up (first to see) and the rest of the list to follow. Having both generic and labeled names in the list is a pattern that is also used in the STM32 boards csv files.

So with the STM32 in mind I would leave the pin order as is.

glennrub replied:
If i recall correctly the duplicate definition would not have any effect. The make-pins.py for nrf-port is a bit different compared to the stm32 port.

The expectation is that, 1. column will generate the name for Pin.board, and the 2. column for Pin.cpu. This means that a definition of 'BTN,P6' should render Pin.board.BTN and Pin.cpu.P6. This will keep both the generic name from the CPU and the unique name for the Board.

What happens for the duplicate is a bit unknown. Does it run on target, can you do the lookup properly?

rolandvs committed:
Updated links to nrfjprog to get latest version

The nrfjprog and mergehex cli are updated to `v9.7.3`

rolandvs committed:
Updated the pins.csv

rolandvs committed:
Cleaned up definitions

Definitions minimized and tested it on the target board.

glennrub commented on ports/nrf/README.md:
```diff
@@ -106,6 +107,7 @@ wt51822_s4at         | s110                    | Peripheral             | Manual
 pca10040             | s132                    | Peripheral and Central | [Segger](#segger-targets)
 feather52            | s132                    | Peripheral and Central | Manual, SWDIO and SWCLK solder points on the bottom side of the board
 arduino_primo        | s132                    | Peripheral and Central | [PyOCD](#pyocdopenocd-targets)
+WRL13990             |                         | Peripheral and Central | [Segger](#segger-targets)
```
it would be great to keep the board name in lower case here to keep some kind of style. Also, i'm very confident that Bluetooth LE will run on this board, so you could point it to the s132 like the other nrf52832 targets as that will be the proper SoftDevice for this chip.

glennrub commented on ports/nrf/boards/WRL13990/README.md:
```diff
@@ -0,0 +1,72 @@
+# Nordic nRF5x micropython
```
I think this page could be a good supplement for describing the board. For me it would made sense to skip this top level nRF5x title, keep the `WRL13990 Board Support` as title. The section below describes what chip is on the board.

glennrub commented on ports/nrf/boards/WRL13990/README.md:
```diff
@@ -0,0 +1,72 @@
+
+## Example
+
+To get a quick overview of the GPIO pins, a module `show_pins().py` was added in the `nrf/examples` directory. All pins are listed (omitted a few here).
```
I'm thinking that this additional _show_pins_ example should probably be a separate commit being added before this description is added. And if accepted, this section should could point to some example that exists in the past. 

Instead of `was added`it would be more in the terms of `use the script show_pins.py` located in the examples/ folder.

rolandvs replied:
I could remove it al together and just list the script in the readme and leave it to the reader to copy and paste it for his own benefit?

glennrub commented:
@rolandvs , i am not sure if it makes sense to update the main.c and add a new example as part of board support. Would it be an idea to split the commit series into two PR's, first adding the board support itself?

The pure board support part of the PR, with a few more modifications (docs), starts looking OK to me with the latest updates on the `pins.csv` and `mpconfigboard.h`.

For the main.c update, i am wondering if the script in main.c,  for listing pins, is worth the flash size it will consume. is this something that all targets would need as inbuilt? I'm not opposing adding a helper script in examples/ if it makes sense, but my opinion is that main.c should be kept small. Even the LED script which is already present is a bit big to me, but kept in order to show where users could inject code for their own use cases when building from source. And, also show that the board is alive after flashing.

rolandvs committed:
Corrected table regarding the wrl13990

rolandvs committed:
wrl13990 readme

rolandvs commented:
@glennrub - I REMOVED the addition to the `main.c` but in a way that this is really *lost*. Apart from history reasons, it is not a problem :-). Also removed the example `show_pins.py` and put it into the board `README.md` file. Changed the `nrf/README.md` so I think "it is ready" :-).

glennrub commented on ports/nrf/boards/WRL13990/README.md:
```diff
@@ -0,0 +1,91 @@
+
+LABEL | BOARD | CPU | REMARK
+--- | --- | --- | ---
+         |P0 |P0 | Reserved 32kHz XTAL 
```
I'm sorry i am nitpicking here... 
I feel the board/cpu columns share a bit redundant information. Does it make sense to document the board (silkscreen label) and the CPU pin as two distinct values if they will always be the same? I'm thinking that it could make sense to maybe combine these two columns. What do you think?

glennrub commented:
As target name now has been aligned to lowercase as requested, the folder board/WRL13990 should also be named in lowercase to align. This makes it possible for the Makefile to find the correct board folder for when issuing `make BOARD=wrl13990`.

rolandvs commented:
Funny I did change the directory by renaming it to the lower case version. I see it in my local repo and use `make BOARD=wrl13990`. So what could be wrong?

rolandvs committed:
Remove some "redundancy" from table

aykevl commented on ports/nrf/boards/WRL13990/mpconfigboard.h:
```diff
@@ -0,0 +1,64 @@
+#define MICROPY_HW_UART1_RX         (26)    // default 115200
+#define MICROPY_HW_UART1_TX         (27)
+//#define MICROPY_HW_UART1_CTS        (7)
+//#define MICROPY_HW_UART1_RTS        (5)
```
With flow control disabled, these pins are unused so the definitions can be removed.

rolandvs replied:
I tend to disagree as it is an useful reminder when adding it after all, so I like to keep these definitions around.

aykevl commented:
> Funny I did change the directory by renaming it to the lower case version. I see it in my local repo and use `make BOARD=wrl13990`. So what could be wrong?

You probably didn't commit/push the changes, see `git status`.
To help the move, you can rename the directory back to uppercase and rename it using `git mv`. That will move the directory and add stage the change at the same time in git.

aykevl commented on ports/nrf/README.md:
```diff
@@ -114,13 +116,13 @@ Install the necessary tools to flash and debug using Segger:
 [JLink Download](https://www.segger.com/downloads/jlink#)
 
-[nrfjprog linux-32bit Download](https://www.nordicsemi.com/eng/nordic/download_resource/52615/16/95882111/97746)
+[nrfjprog linux-32bit Download](https://www.nordicsemi.com/eng/nordic/download_resource/52621/24/48793639/97746)
```
These changes should really have been done in a separate PR, as they're unrelated to this board.

rolandvs replied:
I merely updated the current ones with the new ones, but as this is a manual process pointing to the pages with the overview of the current and previous tools is even better.

aykevl commented on ports/nrf/boards/WRL13990/README.md:
```diff
@@ -0,0 +1,90 @@
+A7       |P31| Alternatively ADC0_IN7
+
+## Example
+To get a quick overview of the GPIO pins, a module `show_pins.py` is in the section below. It gives you an overview of all the defined pins and the way they are configured.
```
While a script to show pins could be useful, I agree with @glennrub that it should be put in the examples/ folder (in a separate pull request). It certainly shouldn't be put in this README as it is not board-specific (except for this particular pin mapping which is already included in this README).

aykevl commented on ports/nrf/boards/WRL13990/README.md:
```diff
@@ -0,0 +1,90 @@
+# wrl13990 Board Support
+
+The Sparkfun `WRL13990` is a very basic, no thrills, `nRF52832-QFAA` based board. It can be used as a DIL socket and eases the development of a Bluetooth Low Energy or other ARM Cortex-M4 applications.
```
Nitpick: this sentence isn't right, it has one "... a ..." too much.

aykevl commented on ports/nrf/boards/WRL13990/README.md:
```diff
@@ -0,0 +1,90 @@
+# wrl13990 Board Support
```
Nitpick: the rest of the document uses uppercase WRL13990 so the title should certainly use that.
Also, the SparkFun website doesn't mention this name, which might be confusing. In other places it appears to be called WRL-13990 (with a dash), so using WRL13990 here may be confusing.

aykevl commented:
Some other comments:

 * It would be nice to link to the board README from the port README. See [this commit](https://github.com/aykevl/micropython/commit/30ce6be612fb52496352e7e391084dd4952a1363) for how it can be done.
 * The tables in the board README are messed up ([see here](https://github.com/rolandvs/micropython/blob/SPARKFUN_NRF52832/ports/nrf/boards/WRL13990/README.md)). For a guide, see: https://help.github.com/articles/organizing-information-with-tables/

rolandvs committed:
nrtf: rename board directory to be in line with the rest

rolandvs committed:
nrf: set link to board wrl13990 in table

rolandvs committed:
nrf: updated document

Correct naming, typo

Table looked ok in MacDown/atom, however removed some unnecessary spaces

rolandvs committed:
nrf: example script to show all pins and configuration

rolandvs committed:
nrf: removed sample code

rolandvs committed:
nrf: removed typo

rolandvs committed:
nrf: typo

some nitpicking myself...

rolandvs commented:
@aykevl @glennrub Should I squash all the above clutter into a single one?

glennrub commented:
@rolandvs I guess a squash could be good in the end. It could also make sense to keep the script addition (examples/show_pins.py) as a separate commit as it is a bit unrelated to the board support itself. 

When that's said, I still have a few questions about the script to list pins. Where is the original source of this? My thinking is that if there is already a script that does this already (which was adopted to nrf), it might be a good idea to look if we can make it common instead of introducing a port specific variant of it.

rolandvs commented:
About the script, @dhylands made it a while ago. There is a new one that is a bit more "pretty" print, which you will find in the `examples/` directory of micropython. The change I made had to do that the `dir(pyb.Pin)` or `dir(machine.Pin)` now also outputs `__class__` and `__name__`  that will break the original script. What it should do is give an overview of all the pins and preferably the selected `af()` functions to get an overview. For now I could drop it from the port so that can be finished off and see if there is a need to add the script again in this form or more advanced...

glennrub commented on ports/nrf/README.md:
```diff
@@ -106,6 +107,7 @@ wt51822_s4at         | s110                    | Peripheral             | Manual
 pca10040             | s132                    | Peripheral and Central | [Segger](#segger-targets)
 feather52            | s132                    | Peripheral and Central | Manual, SWDIO and SWCLK solder points on the bottom side of the board
 arduino_primo        | s132                    | Peripheral and Central | [PyOCD](#pyocdopenocd-targets)
+[wrl13990](ports/wrl13990)| s132                    | Peripheral and Central | [Segger](#segger-targets)
```
A minor syntactic sugar detail, but could you expand the "Target board (BOARD)" with a few spaces, and the also add some blank spaces on the other targets so that the PIPE characters get aligned?

glennrub commented:
> For now I could drop it from the port so that can be finished off and see if there is a need to add the script again in this form or more advanced...

@rolandvs, Sound good to me. If you drop the `show_pins.py` and the reference to it in the `wrl13990/README.md` that will make me happy. We could also raise an Issue to discuss what to do about the `examples/` script. It might trigger some changes related to `machine.Pin` in the nrf-port as well, in order to align the other way around, more towards the already existing script instead of adopting it to work for nrf-port.

glennrub commented on ports/nrf/boards/wrl13990/README.md:
```diff
@@ -0,0 +1,68 @@
+- When UART hardware handshaking is required, select the GPIO's you like and set them accordingly in the `mpconfigboard.h`.
+
+
+BOARD | CPU | REMARK
```
A minor syntactic sugar detail on this table as well. Could you expand the columns so that the PIPE characters get aligned? It makes it a bit easier to read when using a text editor.

rolandvs replied:
Ok that should do it :-)

rolandvs committed:
nrf: neat table, removed unneccessary link

rolandvs committed:
Updated wrl13990 readme

rolandvs committed:
nrf: resolved old links

glennrub commented:
@rolandvs , I realized i made you remove the link from the `nrf/README.md` to the `boards/wrl13990/README.md`. That was not my intention. I believe @aykevl comment is still valid to keep such a link. Given that the link to the board documentation is back, my comment was about aligning the tabbing in the section so that it looks nice to read. And i guess the `show_pins.py` could also be dropped for now as discussed previously. 

Also, @aykevl comment about he updated links to Segger download is a bit unrelated in context of adding a new board and should ideally have been a separate PR. I guess we can squash the commits (70a12750d9de7ae82a5218227253e50234e6fcbe, 4426c0de00aeb791f3215c60cceddb60d6461ece) into a single commit which is not part of the pure board support part of the PR, or simply just remove these changes and create a new PR for these. Right now, to get this board support PR forward i'm leaning towards the latter, simply just drop these Segger link changes for now as well.

To summarize, my suggestions would be:
* Add back the link to the `boards/wrl13990`in the `nrf/README.md`
* Align the | characters in the Target Boards table so that the table looks nice in a text editor
* Drop the show_pins.py script 
* Drop the Segger link updates

Does this sound reasonable?

rolandvs commented:
This PR is closed and replaced by #3991. All suggestions are processed:
- Sample code to show the pin configuration is removed
- The nrfjprog links were updated using a PR #3964
